### PR TITLE
Implement text blinking

### PR DESCRIPTION
**Pull request by yamabiiko**
Implements text blinking when entering the proper escape codes as requested in #2365

### CODE REVIEW

File: alacritty_terminal/src/term/cell.rs [hunk 1]
```diff
@@ -27,6 +27,7 @@ bitflags! {
         const UNDERCURL                 = 0b0001_0000_0000_0000;
         const DOTTED_UNDERLINE          = 0b0010_0000_0000_0000;
         const DASHED_UNDERLINE          = 0b0100_0000_0000_0000;
+        const BLINK                     = 0b1000_0000_0000_0000;
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty_terminal/src/term/cell.rs:30]: Yes, we have 2 bytes more just for that.

**Review by chrisduerr — COMMENTED** [review]
I've left the original ticket open hoping nobody would ever bother. But do we actually want to support this? I can't remember ever thinking to myself that I *want* text to blink. The only example I know of that uses it (for no benefit whatsoever) is wttr.in.

@kchibisov do you have any thoughts on this?

**Comment by nixpulvis** [comment]
If my opinion is worth anything...

I personally feel like applications should use this feature extremely conservatively, and with the understanding that users may have disabled this feature entirely.

I can imagine use cases, and therefor would personally support it's addition, as long as it (like any other animations) has a configuration option to disable.

**Comment by chrisduerr** [comment]
> If my opinion is worth anything...

Of course it is.

> I personally feel like applications should use this feature extremely conservatively, and with the understanding that users may have disabled this feature entirely.

> I can imagine use cases, and therefor would personally support it's addition, as long as it (like any other animations) has a configuration option to disable or change the timing of.

Key point here seems to angle around configuration of the blinking option. Curious what you would suggest as a default.

**Comment by yamabiiko** [comment]
I agree that it would make sense to make it config-based, I can change that.
I think it could be defaulted to enabled considering how it's implemented - the blink event wont be sent if there are no blinking characters anyways.

**Comment by kchibisov** [comment]
>@kchibisov do you have any thoughts on this?

I'm not sure I want anything to blink. I'm not even sure what the use case for that could be. Though, I think `less` is using blinking text by default...

**Comment by chrisduerr** [comment]
> Though, I think less is using blinking text by default...

Do you know for what? I couldn't find anything with `xterm` in `man`.

**Comment by kchibisov** [comment]
>Do you know for what? I couldn't find anything with xterm in man.

I'm not sure, but it's `man less`.

I think you've seen some text blinking in man pages, but it depends how you setup less...

In general I don't want to see blinking text like ever in my terminal since it's just destructing. For cursor it's different since it grabs your attention, so it's easier to find it.

**Comment by chrisduerr** [comment]
@yamabiiko May I ask what you want blinking text for specifically?

**Comment by yamabiiko** [comment]
I think it's a feature that could be useful in certain situations, like as a way to visually notify that a script is done executing or requires user input when having a lot of terminal windows with a tiling desktop. 
But I understand that the use cases are more niche than not.

**Comment by kchibisov** [comment]
>I think it's a feature that could be useful in certain situations, like as a way to visually notify that a script is done executing or requires user input when having a lot of terminal windows with a tiling desktop.
But I understand that the use cases are more niche than not.

I think `bell` and window urgency is what is used for that.

**Comment by chrisduerr** [comment]
> I think it's a feature that could be useful in certain situations, like as a way to visually notify that a script is done executing or requires user input when having a lot of terminal windows with a tiling desktop.
But I understand that the use cases are more niche than not.

So you currently don't have a specific usecase for this either, you'd just like to add it to try it out for yourself?

**Comment by chrisduerr** [comment]
> I don't have a specific usecase, I just used text blinking in the past with other terminal emulators in various scripts.
The main reason I implemented this is because I wanted to contribute to the project and thought it would be something wanted for the sake of feature parity (and because there is an open issue related to it).

Fair enough. For people that are interested in this feature this PR should help them build it in from source. But given that the use of blinking text is somewhat questionable, I think it's better not to merge this into Alacritty at this point unless someone has a convincing argument to do so.

**Comment by nixpulvis** [comment]
I spent about 30 minutes trying to find a video from a movie where there's some alert flashing on a terminal. Like, <blink>M1 ASTROID INCOMING</blink>

**Comment by chrisduerr** [comment]
> P.S. the fact GitHub suppresses the <blink> tag can be submitted as Exhibit A for why this feature might be dubious, perhaps... though I generally like to think people have a bit more discipline over what then run on their terminal.

Yeah, browsers actually removed support for that because it sucks. And while I agree that terminals are much less likely to suffer from it as much as browsers, some features are just better kept dead.

**Comment by j4james** [comment]
A couple of use cases I'm aware of:

1. Blinking is sometimes used for aesthetic effects in ASCII/ANSI animations and games. Think of lights on a Christmas tree, or twinkling stars in the sky. Some examples from YouTube: [Animated Christmas cards](https://youtu.be/wXFjtV6ugVo?t=74), [a BBS game](https://youtu.be/V_os2_RBrfI?t=260s), and [another screen from that game](https://youtu.be/V_os2_RBrfI?t=1459). Admittedly not great examples, but those were just the first couple of links I found.
2. For users with certain forms of visual impairment, colored text can be difficult to differentiate, so it can help to have other choices of styling to highlight important content. One specific example I recall was using blinking text to indicate orphan symlinks in bash (I'm almost certain this used to be the default actually, but I might be misremembering).

I'm not a regular Alacritty user myself, so I'm not arguing for the inclusion of this feature. I just wanted to offer some insight into why some people might appreciate support for blinking, even if it's just an option that is disabled by default.

**Comment by Tony-Sol** [comment]
Is here a usecase of blinking text required? If so, i have another one:
For, me blinking text is nice opportunity to get current tmux mode just using my peripheral vision - without focus on it's indicator; here is a quick demo:
https://user-images.githubusercontent.com/33671815/195414931-220fa10e-c2e3-4a46-9fca-189c44a3e3c1.mov

Of cause, it's extremely rare usecase, but in general - get user attention regardless of time is not
Thats why bell is not an option, because you can't just "make it rings" all the time

**Comment by DCCXX** [comment]
Don't know why exactly this issue (with a PR) got closed, but I just found myself having the same problem.

I'm always irritated by TMUX not working properly, just because this simple function is lacking from the otherwise perfectly functional alacritty.

Where is the danger? I don't why this doesn't get merged and we move on.

**Comment by erikbloodshed** [comment]
Kabobo ng mga core developer neto. Dami rin nalinlang ng mga gumagamit ng app na ito.

**Comment by cpakkala** [comment]
The only arguments against seem to be that blinking is annoying and that can be alleviated by making the feature configurable and setting it to disabled by default.  Is there really concern that this "unneeded" feature is going to take extra resources?  Personally, I like using blinking text for orphaned symlinks...

**Comment by jeeftor** [comment]
I would love to have blinking text implemented. I like to use it in my prompt sometimes

**Comment by bcgraham** [comment]
FWIW, I use blinking text in vim, to highlight the current search term, highlight group `CurSearch` (admittedly abusing `t_Ts`, intended for strikethrough, which I don't use). I also use it in the tmux status bar when `synchronize-panes` is enabled, to keep it in my attention.

**Comment by uhop** [comment]
Blinking text is very effective to attract someones attention. I use it in my scripts when a user action is required, which I cannot (or don't want to) script such as required system modifications or system reboot.

It helps me when I work with a lot of remote computers switching between windows frequently. I have no time patiently re-read text, which does not require my immediate attention.

If some important person doesn't like this feature, it could be enabled/disabled in the configuration file so that important person would be completely unaffected. If, purely hypothetically, someone doesn't like mauve, it is not a reason to prohibit it. Just don't use it personally. Nobody is forcing anyone.

PS: I use alacritty + tmux. I just got it to the "works-for-me" point. I really hate switching and starting the journey anew.

**Comment by LordOfTrident** [comment]
Would be a useful feature, i use it in my TUI stuff and decoration scripts which was already mentioned as a use case before, i was honestly kinda shocked to find out such a big terminal project does not support the blinking attribute.

**Comment by ryanmjacobs** [comment]
I know this is years later... but I also find this useful.

**Comment by carlanwray** [comment]
New to Alacritty and I'm trying to understand the lack of intellectual rigor in this situation. 

In my case I want the status line in Claude Code to blink CONTEXT in red when I reach x number of tokens. And while I don't have any netops TUI's currently there are use cases there as well.

All of that aside, just because an individual doesn't have a use case, clearly someone wanted it bad enough to issue a pull request for something relatively simple, that, while an edge case, is still relevant.

Using GitHub not allowing <blink> as a case against this is absurd at best, they are controlling an issue, not arbitrarily limiting functionality.

/old-man-shouting-into-the-wind

**Pull request closed without merging**